### ISSUE TITLE
T5 mask bug when calculate loss.

### ISSUE THREAD

**Issue by wlhgtc** [allenai/allennlp#5342]
The CrossEntropy loss used in T5 model would ignore pos where `idx = -100`.

According to the [code](https://github.com/huggingface/transformers/blob/master/src/transformers/data/data_collator.py#L277-L284) in transformers, they set all padding tokens to -100. 
In our code we miss this operation, so we calculate these loss which is unnecessary and would lead a little bit drop(at least in my own task).

Such code should be added:
```python
labels.masked_fill_(labels == 0, -100)
```
Add it in  [allennlp](https://github.com/allenai/allennlp/blob/main/allennlp/modules/transformer/t5.py#L930) or [allennlp-models](https://github.com/allenai/allennlp-models/blob/main/allennlp_models/generation/models/t5.py#L101) ?

**Comment by wlhgtc** [comment]
@dirkgr @AkshitaB

**Comment by dirkgr** [comment]
@AkshitaB, are you on top of this? I think padding correctly is becoming one of the more complicated things we do in AllenNLP 🤪.

**Comment by AkshitaB** [comment]
@wlhgtc @dirkgr I am looking into this. I don't quite understand why our tests that compare HF loss with our loss don't fail.

**Comment by wlhgtc** [comment]
> @wlhgtc @dirkgr I am looking into this. I don't quite understand why our tests that compare HF loss with our loss don't fail.

It works all the same when `batch == 1` that means no mask. And if you show me the test code, I'm willing to help.

**Comment by AkshitaB** [comment]
@wlhgtc Here's our test that compares to HF: https://github.com/allenai/allennlp/blob/main/tests/modules/transformer/t5_test.py#L45 

The batch size is 2.

**Comment by wlhgtc** [comment]
> @wlhgtc Here's our test that compares to HF: https://github.com/allenai/allennlp/blob/main/tests/modules/transformer/t5_test.py#L45
> 
> The batch size is 2.
You miss some preprocess operation(the `DataCollatorForSeq2Seq `) when hf model.
This [method](https://github.com/huggingface/transformers/blob/master/src/transformers/data/data_collator.py#L237) was call before feed input to model , mask all padding tokens to -100. 
But you feed the same inputs to the two models.

**Comment by dirkgr** [comment]
I think I found a bug in that HF code snippet you posted: https://github.com/huggingface/transformers/issues/13341

But that has nothing to do with this issue. I'm still investigating what the right thing to do is here.

**Comment by dirkgr** [comment]
It seems to me that the way to solve this in AllenNLP is to set the padding value for your labels to `-100`. We should make this more discoverable though, maybe even make it a default somewhere. @wlhgtc, can you share a few more details about your code? Where/how are you using this?

I don't want to just set all `0` to `-100`, because that's not the right thing in some cases (and even in HF, this is behavior you can turn on or off).

**Comment by wlhgtc** [comment]
@dirkgr 
I use the code in [allennlp-models](https://github.com/allenai/allennlp-models/blob/main/allennlp_models/generation/models/t5.py#L104), just one line diff
```python
if target_tokens is not None:
    labels, decoder_attention_mask = (
        target_tokens["tokens"]["token_ids"],  # type: ignore[assignment]
        target_tokens["tokens"]["mask"],  # type: ignore[assignment]
    )
    labels.masked_fill_(labels == 0, -100)  #  set padding to -100
elif self.training:
    raise ValueError("'target_tokens' required during training")
```
But there is still another way, we could modify our loss function in [here:](https://github.com/allenai/allennlp/blob/main/allennlp/modules/transformer/t5.py#L799)
```python
# self.loss_fct = CrossEntropyLoss(ignore_index=-100)
self.loss_fct = CrossEntropyLoss(ignore_index=PAD_INDEX)
```

**Comment by dirkgr** [comment]
In principle I like the second one better, but worry about breaking the pact that we produce the same numbers that HF produces given the same input. I don't think we can safely break that assumption.

You must have some field that passes in your labels, right? Why can't you set the padding index on that field?
